File: Data Structures & Algorithms/permutations-ii/submission-0.java
class Solution {
    private List<List<Integer>> res;

    public List<List<Integer>> permuteUnique(int[] nums) {
        res = new ArrayList<>();
        backtrack(nums, new boolean[nums.length], new ArrayList<>());

        return res;
    }

    private void backtrack(int[] nums, boolean[] picked, List<Integer> perm){
        if (perm.size() == nums.length){
            if (!res.contains(perm)){
                res.add(new ArrayList(perm));
            }
            return;
        }

        for (int i = 0; i < nums.length; i++){
            if (!picked[i]){
                picked[i] = true;
                perm.add(nums[i]);
                backtrack(nums, picked, perm);
                picked[i] = false;
                perm.remove(perm.size() - 1);
            }
        }
    }
}

File: Data Structures & Algorithms/permutations/submission-0.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {number[][]}
     */
    permute(nums) {
        const res = [];

        const backtrack = (perm, nums, pick) => {
            if (perm.length === nums.length){
                res.push([...perm]);
                return;
            }

            for (let i = 0 ; i < nums.length; i++){
                if (!pick[i]){
                    perm.push(nums[i]);
                    pick[i] = true;

                    backtrack(perm, nums, pick);

                    perm.pop();
                    pick[i]  = false;
                }
            }
        }

        backtrack([], nums, new Array(nums.length).fill(false));

        return res;
    }
}
